Register CIA ops to FakeTensorMode directly in export

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140465

During export, we nub out most CIA ops to return NotImplemented to avoid decomposing them during tracing. To recover the existing shape propagation behavior, we register these CIA decomps directly as FakeTensorMode rules as well. The reason we have to do is because when we return NotImplemented, FakeTensor would fallback to running these CIAs with Meta backend causing device branching CIA ops to fail. (because now the device is Meta. One example is sdpa). If we register a kernel directly to FakeTensorMode, we won't fallback to Meta backend. 


Differential Revision: [D65716260](https://our.internmc.facebook.com/intern/diff/D65716260/)

tugsbayasgalan committed:
Move on_device_ai unittests to training IR

Export team is moving to non-functional training IR (See more details at: https://fb.workplace.com/groups/257735836456307/permalink/746346547595231/)

We expect this change to be theoretically no-op for executorch users as the to_edge call where we lower to backends should normalize the IR to inference specific. In this diff, I change some on_device_ai use cases to reflect above IR change. If this change goes through well, i will roll it out to all on_device_ai use cases instead of just unittests.

Differential Revision: [D65716260](https://our.internmc.facebook.com/intern/diff/D65716260/)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Move on_device_ai unittests to training IR"


Export team is moving to non-functional training IR (See more details at: https://fb.workplace.com/groups/257735836456307/permalink/746346547595231/)

We expect this change to be theoretically no-op for executorch users as the to_edge call where we lower to backends should normalize the IR to inference specific. In this diff, I change some on_device_ai use cases to reflect above IR change. If this change goes through well, i will roll it out to all on_device_ai use cases instead of just unittests.

Differential Revision: [D65716260](https://our.internmc.facebook.com/intern/diff/D65716260/)

[ghstack-poisoned]

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -224,6 +229,31 @@ def _override_composite_implicit_decomp(cia_ops_to_callable, safe=True):
+        # Note that is a no-op in most cases, because:
+        #   1) In post dispatch tracing, CIA would have already decomposed
+        #   2) Most CIA impl are device agnostic.
+        def _force_dispatch_to_orig_cia_callable(fake_tensor_mode, op, *args, **kwargs):
```
nit: you're defining this local fn in a giant for loop that loops over every CIA op: maybe define the fn itself outside of the loop?

tugsbayasgalan replied:
Oh lol yeah sorry my bad

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -250,6 +280,7 @@ def _override_composite_implicit_decomp(cia_ops_to_callable, safe=True):
             op.py_kernels.clear()
             op.py_kernels.update(saved_tables[op])
             op._dispatch_cache.clear()
+            _deregister_op_impl(op)
```
this seems wrong: above, you only registered a fake rule for ops that didn't have an existing fake rule. But here, you're *unconditionally** deregistering fake rules for all CIA ops (even if they had an existing fake rule prior to running export).

You probably want to keep a list of which ops you actually registered fake rules for, and only deregister those?

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -224,6 +229,31 @@ def _override_composite_implicit_decomp(cia_ops_to_callable, safe=True):
                 decomp_callable
             )
 
+        # [NOTE] Directly registering fake tensor rule to CIA ops
```
maybe just a nit that's helpful to me, but I think it would be useful to mention something like:

There are effectively 3 places where an op register tensor propagation logic:

(1) as a CompositeImplicitAutograd decomposition (above FakeTensorMode)
(2) as a FakeTensorMode registration
(3) as a `DispatchKey::Meta` registration

During export, we nub out most CIA ops to return `NotImplemented` to avoid decomposing them during tracing. To recover the existing shape propagation behavior, we register these CIA decomps directly as FakeTensorMode rules as well.

A valid question is: all CIA ops are also registered to the Meta key, so why can't we just rely on running the meta tensor kernels? The issue is that when FakeTensorMode runs a meta tensor kernel, the devices of all inputs have been nubbed out to return "meta". If we have any CIA op impls that need to branch on device, we need to register them as FakeTensorMode rules to preserve any device-specific behavior.

tugsbayasgalan committed:
Update on "Fix scaled_dot_product_issue"



Differential Revision: [D65716260](https://our.internmc.facebook.com/intern/diff/D65716260/)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix scaled_dot_product_issue"



Differential Revision: [D65716260](https://our.internmc.facebook.com/intern/diff/D65716260/)

[ghstack-poisoned]

bdhirsh commented:
Also - can you change the PR title/description? (the issue isnt' really specific to sdpa)

tugsbayasgalan committed:
Update on "Register CIA ops to FakeTensorMode directly in export "


During export, we nub out most CIA ops to return NotImplemented to avoid decomposing them during tracing. To recover the existing shape propagation behavior, we register these CIA decomps directly as FakeTensorMode rules as well. The reason we have to do is because when we return NotImplemented, FakeTensor would fallback to running these CIAs with Meta backend causing device branching CIA ops to fail. (because now the device is Meta. One example is sdpa). If we register a kernel directly to FakeTensorMode, we won't fallback to Meta backend. 


Differential Revision: [D65716260](https://our.internmc.facebook.com/intern/diff/D65716260/)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Register CIA ops to FakeTensorMode directly in export "


During export, we nub out most CIA ops to return NotImplemented to avoid decomposing them during tracing. To recover the existing shape propagation behavior, we register these CIA decomps directly as FakeTensorMode rules as well. The reason we have to do is because when we return NotImplemented, FakeTensor would fallback to running these CIAs with Meta backend causing device branching CIA ops to fail. (because now the device is Meta. One example is sdpa). If we register a kernel directly to FakeTensorMode, we won't fallback to Meta backend. 


Differential Revision: [D65716260](https://our.internmc.facebook.com/intern/diff/D65716260/)

[ghstack-poisoned]

bdhirsh commented on test/export/test_export.py:
```diff
@@ -9026,6 +9026,28 @@ def test_dynamic_shapes_serdes_user_errors(self):
+                return x
+
+        inps = (torch.randn(1, 224, 768, device="cpu"),)
+        export(Foo(), inps)
```
would this test have failed prior to this change? (IIUC, the main change in this test is that it will "fix" any CIA decomps that branch on cpu/cuda device, because previously we would execute them with meta tensors)

tugsbayasgalan replied:
Added a comment explaining the old behaviour

tugsbayasgalan committed:
Update on "Register CIA ops to FakeTensorMode directly in export "


During export, we nub out most CIA ops to return NotImplemented to avoid decomposing them during tracing. To recover the existing shape propagation behavior, we register these CIA decomps directly as FakeTensorMode rules as well. The reason we have to do is because when we return NotImplemented, FakeTensor would fallback to running these CIAs with Meta backend causing device branching CIA ops to fail. (because now the device is Meta. One example is sdpa). If we register a kernel directly to FakeTensorMode, we won't fallback to Meta backend. 


Differential Revision: [D65716260](https://our.internmc.facebook.com/intern/diff/D65716260/)

[ghstack-poisoned]
